### PR TITLE
Graph break overhead

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/TRTEngine.cpp [hunk 1]
```diff
@@ -281,6 +289,14 @@ void TRTEngine::enable_profiling() {
   exec_ctx->setProfiler(trt_engine_profiler.get());
 }
 
+void TRTEngine::set_unowned_output_tensor(bool enable) {
+  this->unowned_output_tensor = enable;
+}
+
+bool TRTEngine::is_unowned_output_tensor() {
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp]: Can we adjust these names so they are clear: 

```
TRTEngine::set_output_tensors_as_unowned(bool enable);
TRTEngine::are_output_tensors_unowned(bool enable):
```

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 2]
```diff
@@ -218,9 +219,30 @@ def __init__(
         self.requires_output_allocator = requires_output_allocator
         self.output_allocator: Optional[DynamicOutputAllocator] = None
         self.use_output_allocator_outputs = False
-
+        self.device = torch.cuda.current_device()
+        self.cudagraphs_enabled = torch_tensorrt.runtime.get_cudagraphs_mode()
+        # If the output tensor is not owned by the engine (unowned_output_tensor=True), we need to create a new output tensor in each forward pass
+        self.unowned_output_tensor = False
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: Same here `self.output_tensors_are_unowned`

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 3]
```diff
@@ -218,9 +219,30 @@ def __init__(
...
         if self.serialized_engine is not None and not self.settings.lazy_engine_init:
             self.setup_engine()
+        self.is_shape_inference_io = {
+            input_name: self.engine.is_shape_inference_io(input_name)
+            for input_name in self.input_names
+        }
+
+    def set_unowned_output_tensor(self, enabled: bool) -> None:
+        """
+        Set the flag to indicate if the output tensor is unowned by the engine.
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: To clarify the description here: 

```
Flag to set if the output tensors of this engine are wholy owned by the Torch-TensorRT Runtime or if they might be shared with a user. 

If the tensors are not owned by the runtime, then they must be recreated on every forward call which may have implications for performance.

Typically only the final engine in a graph requires output tensors to be unowned and there are performance gains to be had for intermediate engines to manage their own standing memory. 

Therefore this should only be set to True for the final module in a graph and leave false for intermediate modules.

```

File: core/runtime/execute_engine.cpp [hunk 4]
```diff
@@ -145,10 +146,10 @@ void setup_input_tensors(
         // Create a new persistent input buffer
         compiled_engine->input_buffers[i] = std::move(formatted_inputs.back().clone());
       }
-
-      TORCHTRT_CHECK(
-          compiled_engine->exec_ctx->setInputShape(name.c_str(), dims), "Error while setting the input shape");
-
+      if (shape_changed) {
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp]: Can't we use the same shape keys we used for cudagraphs? Why are we implementing another system?

File: core/runtime/TRTEngine.cpp [hunk 5]
```diff
@@ -237,6 +237,14 @@ TRTEngine::TRTEngine(
       out_binding_names[pyt_idx] = binding_name;
     }
     num_io = std::make_pair(inputs_size, outputs);
+
+    this->current_device_id = at::cuda::current_device();
+    this->stream = c10::cuda::getCurrentCUDAStream(this->current_device_id);
```
Comment (narendasan) [inline core/runtime/TRTEngine.cpp]: Is this PR supposed to implement the stream operator?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 6]
```diff
@@ -949,6 +950,7 @@ def preserve_module_specs(
     for attr in dir(gm):
         if attr.startswith("_frozen_param"):
             delattr(gm, attr)
+    trt_module = None
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: What is this? Why does this variable need to be outside the of scope of iterating through the module?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 7]
```diff
@@ -1070,20 +1072,28 @@ def preserve_module_specs(
...
 
     # Replace all FX Modules with TRT Modules
     for name, trt_module in trt_modules.items():
         setattr(partitioned_module, name, trt_module)
         if settings.lazy_engine_init and not settings.enable_cross_compile_for_windows:
-            getattr(partitioned_module, name).setup_engine()
+            trt_module = getattr(partitioned_module, name)
+            trt_module.setup_engine()
+
+    if trt_module:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: I feel like this is a pretty round about way to get the last engine

Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py]: Also the order from items is not deterministic so I dont think this is guaranteed to give you the last module anyway. You might need to work back from outputs.

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 8]
```diff
@@ -1070,20 +1072,28 @@ def preserve_module_specs(
...
+        trt_module.set_output_tensors_as_unowned(True)
 
     # Reset settings object to user specification after fallback to global partitioning mode
     if fast_partitioner_failed:
         settings.use_fast_partitioner = True
 
     dryrun_stats_display(dryrun_tracker, settings.dryrun)
+    if not settings.dryrun:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Why is this conditioned on dry run? Dry run should reflect the deployment graph

File: setup.py [hunk 9]
```diff
@@ -195,10 +195,10 @@ def build_libtorchtrt_cxx11_abi(
     else:
         cmd.append("//:libtorchtrt")
 
-    if develop:
-        cmd.append("--compilation_mode=dbg")
-    else:
-        cmd.append("--compilation_mode=opt")
+    # if develop:
```
Comment (narendasan) [inline setup.py]: Remove

File: py/torch_tensorrt/dynamo/runtime/_stream_handler.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                    torch.ops.tensorrt.exit_compute_stream
+                )
+
+    partitioned_module.graph.lint()
+    partitioned_module.recompile()
+    return partitioned_module
+
+
+@torch.library.custom_op("tensorrt::enter_compute_stream", mutates_args=())
+def enter_compute_stream() -> None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_stream_handler.py]: IMO these should likely be dependent on the input tensors and pull their stream information from there. You arent necessarily guaranteed for the input tensors to be on the default stream. It also means you need storage for the streams. 


Also how does this work in c++?

File: py/torch_tensorrt/dynamo/runtime/_stream_handler.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import torch
+import torch.fx
+
+
+def handle_cuda_stream(
+    partitioned_module: torch.fx.GraphModule,
+) -> torch.fx.GraphModule:
+    for node in partitioned_module.graph.nodes:
+        if node.op == "placeholder":
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_stream_handler.py]: what if there are multiple placeholders or outputs?

File: core/runtime/execute_engine.cpp [hunk 12]
```diff
@@ -140,15 +140,17 @@ void setup_input_tensors(
     } else {
       at::Tensor contig_input = inputs[i].view(shape).contiguous();
       formatted_inputs.emplace_back(std::move(contig_input));
-
+      bool need_cudagraphs_record = cudagraphs_enabled and
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp]: Just use &&, ||, !

File: core/runtime/execute_engine.cpp [hunk 13]
```diff
@@ -140,15 +140,17 @@ void setup_input_tensors(
...
+           compiled_engine->runtime_states.context_changed);
       if (need_cudagraphs_record) {
         // Create a new persistent input buffer
         compiled_engine->input_buffers[i] = std::move(formatted_inputs.back().clone());
       }
-
-      TORCHTRT_CHECK(
-          compiled_engine->exec_ctx->setInputShape(name.c_str(), dims), "Error while setting the input shape");
-
+      if (shape_changed or compiled_engine->allocated_outputs.size() == 0) {
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp]: What is the allocated_outputs size test? Just that we havent allocated the outputs yet?

Comment (narendasan) [reply core/runtime/execute_engine.cpp]: Can we fold some of these conditions ahead of time?

Comment (cehongwang) [reply core/runtime/execute_engine.cpp]: So it's like for the first time you run the model or when the input shape get changed, we need to (re)set the input shape right? I don't think we can fold it

File: core/runtime/execute_engine.cpp [hunk 14]
```diff
@@ -248,7 +247,11 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
       if (can_use_pre_allocated_outputs) {
         outputs = compiled_engine->pre_allocated_outputs;
       } else {
-        outputs = create_output_tensors(compiled_engine);
+        if (compiled_engine->allocated_outputs.size() == 0 or compiled_engine->output_tensors_are_unowned or
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp]: Use || instead of 'or'

File: core/runtime/execute_engine.cpp [hunk 15]
```diff
@@ -226,10 +226,9 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
         input_profiler_guard =
             std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->input_profile_path);
       }
-
-      setup_input_tensors(inputs, compiled_engine, cudagraphs_enabled, need_cudagraphs_record);
+      setup_input_tensors(inputs, compiled_engine, cudagraphs_enabled, shape_changed);
       // Check if input shapes can be inferred.
-      int32_t const io_size{compiled_engine->cuda_engine->getNbIOTensors()};
+      int32_t const io_size{compiled_engine->io_size};
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp]: Are things like this actually faster? Like have we looked at the implementation to see its not cached at the TRT level?

File: core/runtime/execute_engine.cpp [hunk 16]
```diff
@@ -248,7 +247,11 @@ std::vector<at::Tensor> execute_engine(std::vector<at::Tensor> inputs, c10::intr
       if (can_use_pre_allocated_outputs) {
         outputs = compiled_engine->pre_allocated_outputs;
```
Comment (narendasan) [inline core/runtime/execute_engine.cpp:249]: What is the difference between "allocated_outputs" and "pre_allocated_outputs"?

File: core/runtime/TRTEngine.h [hunk 17]
```diff
@@ -176,6 +181,7 @@ struct TRTEngine : torch::CustomClassHolder {
   std::string shape_key = "None";
   bool use_pre_allocated_outputs = false;
   std::vector<at::Tensor> pre_allocated_outputs;
+  std::vector<at::Tensor> allocated_outputs;
```
Comment (narendasan) [inline core/runtime/TRTEngine.h]: Is there a way to reuse pre_allocated_outputs?

File: core/runtime/execute_engine.cpp (hunk 15 above)
Comment (cehongwang) [reply core/runtime/execute_engine.cpp]: Yes they do have small difference. I am trying to minize repeating code and this repeatly call this outbound function in every iteration

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:1]: It seems like you are just renaming variables here? Why do we need to make that change?

File: core/runtime/TRTEngine.h (hunk 17 above)
Comment (narendasan) [reply core/runtime/TRTEngine.h]: @cehongwang we need to figure out how to merge these two implementations, we cant just keep adding new code paths for every situation. Either the output allocator needs to change to leverage the new system or the new system needs to leverage what the output allocator uses

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py
Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:1]: the `pre_allocated_outputs` -> `allocated_outputs` change

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 18]
```diff
@@ -480,17 +502,24 @@ def run_standard_execution() -> torch.Tensor | Tuple[torch.Tensor, ...]:
                 else nullcontext()
             ):
                 if can_use_pre_allocated_outputs:
-                    outputs = self.pre_allocated_outputs
+                    outputs = self.allocated_outputs
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: e.g. ^

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:1]: I also changed `self.output_tensors = self.create_output_tensors()` to `self.allocated_outputs = self.create_output_tensors()`. Basically merged two variable together. Isn't that what you wanted me to do?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 19]
```diff
@@ -1070,14 +1070,24 @@ def preserve_module_specs(
...
 
     # Replace all FX Modules with TRT Modules
     for name, trt_module in trt_modules.items():
         setattr(partitioned_module, name, trt_module)
         if settings.lazy_engine_init and not settings.enable_cross_compile_for_windows:
-            getattr(partitioned_module, name).setup_engine()
+            trt_module = getattr(partitioned_module, name)
+            trt_module.setup_engine()
+
+    output_node = list(partitioned_module.graph.nodes)[-1]
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:1085]: Is this ordering gaurenteed, wouldnt it be more correct backtracing from graph outputs?
